homeassistant/components/googlehome/sensor.py

```diff
@@ -0,0 +1,110 @@
+"""
+Support for Google Home alarm sensor.
+
+For more details about this platform, please refer to the documentation at
+https://home-assistant.io/components/sensor.googlehome/
+"""
+import logging
+from datetime import timedelta
+
+from homeassistant.components.sensor import ENTITY_ID_FORMAT
+from homeassistant.components.googlehome import (
+    CLIENT, DOMAIN as GOOGLEHOME_DOMAIN, NAME)
+from homeassistant.const import DEVICE_CLASS_TIMESTAMP
+from homeassistant.helpers.entity import Entity, async_generate_entity_id
+import homeassistant.util.dt as dt_util
+
+
+DEPENDENCIES = ['googlehome']
+
+SCAN_INTERVAL = timedelta(seconds=10)
+
+_LOGGER = logging.getLogger(__name__)
+
+ICON = 'mdi:alarm'
+
+SENSOR_TYPES = {
+    'timer': "Timer",
+    'alarm': "Alarm",
+}
+
+
+async def async_setup_platform(hass, config,
+                               async_add_entities, discovery_info=None):
+    """Set up the googlehome sensor platform."""
+    if discovery_info is None:
+        _LOGGER.warning(
+            "To use this you need to configure the 'googlehome' component")
+
+    devices = []
+    for condition in SENSOR_TYPES:
+        device = GoogleHomeAlarm(hass.data[CLIENT], condition,
+                                 discovery_info)
+        devices.append(device)
+
+    async_add_entities(devices, True)
+
+
+class GoogleHomeAlarm(Entity):
+    """Representation of a GoogleHomeAlarm."""
+
+    def __init__(self, client, condition, config):
+        """Initialize the GoogleHomeAlarm sensor."""
+        self._host = config['host']
+        self._client = client
+        self._condition = condition
+        self._name = None
+        self._state = None
+        self._available = True
+
+    async def async_added_to_hass(self):
+        """Subscribe GoogleHome events."""
+        await self._client.update_info(self._host)
+        data = self.hass.data[GOOGLEHOME_DOMAIN][self._host]
+        info = data.get('info', {})
+        if info is None:
+            return
+        self._name = "{} {}".format(info.get('name', NAME),
+                                    SENSOR_TYPES[self._condition])
+        self.entity_id = async_generate_entity_id(
```

Comment:
Generating the entity_id at this point is too late, if we want to avoid overriding the system set id, or make sure to avoid id clashes by setting the id specifically.

At the latest it has to be done in `update`.

Reply:
We don't seem to need to set the entity_id, either, since we base it on name. So instead just make sure we set name in update (if not set).

Reply:
@MartinHjelmare if I don't generate the entity id id doesn't get set. And defaults to unknown.

Reply:
That's because we don't set name until `async_added_to_hass`. Do that in `async_update` instead, if name is `None`.

Reply:
This should fix that https://github.com/home-assistant/home-assistant/pull/20742 @MartinHjelmare